active/0051-strategies.md

```diff
@@ -0,0 +1,247 @@
+
+Most importantly, the approach described below helps us achieve:
+
++ defining and developing a common language around the subject
++ removing highly coupled code and streamline technical implementation (Ember Engines and Fastboot)
++ unlock a whole different set of plugins we couldn't have before:
+  + ability to create custom bundles (i.e per-engine and per-route bundles)
+  + take advantage of [HTTP2 multiplexing](https://http2.github.io/faq/#why-is-http2-multiplexed) and [cache pushing](https://www.mnot.net/blog/2014/01/30/http2_expectations#4-cache-pushing)
+  + optimising plugins (Javascript and CSS treeshaking)
+
+# Scope
+
++ New public API for registering strategies
+
+# Terminology
+
++ **Strategy** - A strategy is responsible for returning a build pipeline that can emit a specific set of output assets given an input tree.
++ **Assembler** - Responsible for taking the resolved tree and applying – default or user provided – strategies to the tree.
+
+# Detailed design
+
+The detailed design is separated in various sections so that it is easier for a reader to understand.
+
+## Assembler
+
+`Assembler` is similar to a car assembly line - starting with a metal frame, making changes to it along the way, adding parts and at the end, there's an assembled car.
+
+It does not make any assumptions about how the final output should be constructed. Instead, it relies on strategies to tell it what the final output should look like. This is rather important as it enforces clear separation of concerns.
+
+Assembler must have the following interface:
+
+```typescript
+interface Assembler {
+  constructor(inputTree: BroccoliTree, strategies: Strategy[]);
+  toTree(): BroccoliTree;
+}
+```
+
+`toTree` method would gather all the trees from the passed in `strategies`, and merge them into one final broccoli tree.
+
+## Strategies
+
+Strategy is an extensibility primitive. It gives you granular control over the final output. It could be used in many different ways (we are going to go over use cases below).
+
+Strategy must have the following interface:
+
+```typescript
+interface Strategy {
+  constructor(options: any);
+  toTree(assember: Assembler, inputTree: BroccoliTree): BroccoliTree;
+}
+```
+
+`inputTree` will have the following structure:
+
+```sh
+bundler:js:input/
+├── addon-modules
+├── the-app-name-folder
+├── node_modules
+└── vendor
+```
+
+where `addon-modules` contains dependencies from Ember add-ons; `the-app-name-folder` contains Ember application code; `node_modules` contains node dependencies and `vendor` contains other dependencies.
+
+Note, that  `toTree` method must return a broccoli tree.
+
+### Concatenation Strategy
+
+One of the strategies that we already use now to produce `assets/app-name.js` and `assets/vendor.js`. It takes a broccoli tree and returns a new concatenated broccoli tree.
```

Comment:
First sentence here needs to be restructured. You might want to make it clear that although you are proposing Strategies as a public API, a similar architecture is already used internally in Ember-CLI. I think that is implied in your statement.